common/pango.c

```diff
@@ -0,0 +1,67 @@
+#include <cairo/cairo.h>
+#include <pango/pangocairo.h>
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+PangoLayout *get_pango_layout(cairo_t *cairo, const char *font,
+		const char *text, int32_t scale, bool markup) {
+	PangoLayout *layout = pango_cairo_create_layout(cairo);
+	PangoAttrList *attrs;
+	if (markup) {
+		char *buf;
+		pango_parse_markup(text, -1, 0, &attrs, &buf, NULL, NULL);
+		pango_layout_set_markup(layout, buf, -1);
+		free(buf);
+	} else {
+		attrs = pango_attr_list_new();
+		pango_layout_set_text(layout, text, -1);
+	}
+	pango_attr_list_insert(attrs, pango_attr_scale_new(scale));
+	PangoFontDescription *desc = pango_font_description_from_string(font);
+	pango_layout_set_font_description(layout, desc);
+	pango_layout_set_single_paragraph_mode(layout, 1);
+	pango_layout_set_attributes(layout, attrs);
+	pango_attr_list_unref(attrs);
+	pango_font_description_free(desc);
+	return layout;
+}
+
+void get_text_size(cairo_t *cairo, const char *font, int *width, int *height,
+		int32_t scale, bool markup, const char *fmt, ...) {
+	char *buf = malloc(2048);
```

Comment:
Put this on the stack

Reply:
Could also just statically allocate it

Reply:
If you wanted to be "careful" with stack space, you can use `vsnprintf(NULL, 0, fmt, args)` to get the number of characters to fit the string and use a VLA.